### PR TITLE
Feature/packetized data

### PR DESCRIPTION
**Pull request by JakubFara**
## **Summary**

This commit adds support for reading packetized data.
Currently, this functionality is available **only for RVC2 devices**.

---

## **New Structure**

* **`PacketizedData : Buffer`** — sent before the actual packetized data to provide context for subsequent packets.

---

### **`XLinkInHost`**

* **`std::shared_ptr<ADatatype> readData(const XLinkStream& stream) const;`**
  Reads incoming data from the specified stream.

* **`std::shared_ptr<ADatatype> parsePacketizedData(const std::shared_ptr<PacketizedData>& packetizedData, const XLinkStream& stream) const;`**
  Parses packetized data and reconstructs the original payload.

* **`void parseMessageGroup(const std::shared_ptr<MessageGroup>& messageGroup, const XLinkStream& stream) const;`**
  Handles message groups and their contained data packets.

---

### **`XLinkOutProperties`**

* **`int packetSize`** — defines the packet size.
* **`int packetFrequency`** — defines how frequently packets are sent.

---

## **Tests**

* Tests are added to **`tests/src/ondevice_tests/xlink_test.cpp`**.
* The tests verify the expected sending duration base on number of packets and packet frequency.
* Deeper testing is not currently possible, as the public API does not expose low-level packetization details.

### CODE REVIEW

File: include/depthai/pipeline/node/internal/XLinkOut.hpp [hunk 1]
```diff
@@ -49,13 +49,21 @@ class XLinkOut : public DeviceNodeCRTP<DeviceNode, XLinkOut, XLinkOutProperties>
      */
     void setMetadataOnly(bool metadataOnly);
 
+    void setPacketSize(int packetSize);
+
+    void setPacketFrequency(int packetFrequency);
+
```
Comment (moratom) [inline include/depthai/pipeline/node/internal/XLinkOut.hpp:55]: We should add comments here;
* I assume packetSize is in bytes?
* What does packetFrequency do? Is this the "max" frequency?

File: src/pipeline/datatype/PacketizedData.hpp
Comment (moratom) [inline src/pipeline/datatype/PacketizedData.hpp:1]: We should make this be internal (ideally at least)

File: include/depthai/properties/internal/XLinkOutProperties.hpp [hunk 2]
```diff
@@ -18,16 +18,25 @@ struct XLinkOutProperties : PropertiesSerializable<Properties, XLinkOutPropertie
...
 
+    /**
+     * Max bytes per packet (-1 = unlimited)
+     */
+    int packetSize = -1;  //
+
+    /**
+     * Maximal frequency beteen of packets (-1 = unlimited)
+     */
+    int packetFrequency = -1;  //
```
Comment (moratom) [inline include/depthai/properties/internal/XLinkOutProperties.hpp]: Worth considering a max bandiwidth setting instead may be better (easier to grasp and can also be applied in case the messages are not packatized).

File: include/depthai/xlink/XLinkStream.hpp
Comment (moratom) [inline include/depthai/xlink/XLinkStream.hpp:1]: Nice!

**Review by moratom — COMMENTED** [review]
Thanks, looks good!

We should bump RVC4 and RVC2 FW so it compiles and at add a warning on RVC4 that these are not implement at least (or implement the functionality on RVC4)

File: src/pipeline/node/internal/XLinkInHost.cpp [hunk 3]
```diff
@@ -32,6 +33,42 @@ void XLinkInHost::disconnect() {
...
+        parseMessageGroup(messageGroup, stream);
+    }
+    if(auto packetizedData = std::dynamic_pointer_cast<PacketizedData>(msg)) {
+        return parsePacketizedData(packetizedData, stream);
+    }
+    return msg;
+}
+
+std::shared_ptr<ADatatype> XLinkInHost::parsePacketizedData(const std::shared_ptr<PacketizedData>& packetizedData, const XLinkStream& stream) {
+    std::vector<std::uint8_t> payload;
```
Comment (aljazkonec1) [inline src/pipeline/node/internal/XLinkInHost.cpp:53]: Could we reserve the memory for payload beforehand?

File: src/pipeline/node/internal/XLinkInHost.cpp [hunk 4]
```diff
@@ -32,6 +33,42 @@ void XLinkInHost::disconnect() {
...
+
+    for(std::uint32_t i = 0; i < packetizedData->numPackets; ++i) {
+        auto packet = stream.readMove();
+        if(packet.length == 0) {
+            continue;
+        }
+        payload.insert(payload.end(), packet.data, packet.data + packet.length);
+    }
+
+    streamPacketDesc_t packet;
```
Comment (aljazkonec1) [inline src/pipeline/node/internal/XLinkInHost.cpp:73]: Left uninitialized,
```suggestion
    streamPacketDesc_t packet{};
```

File: src/xlink/XLinkStream.cpp [hunk 5]
```diff
@@ -153,12 +153,12 @@ void XLinkStream::read(std::vector<std::uint8_t>& data, XLinkTimespec& timestamp
     }
 }
```
Comment (aljazkonec1) [inline src/xlink/XLinkStream.cpp:154]: This is not part of the changes, but from docstring of `.insert` :
```This function will insert copies of the data in the range [__first,__last) into the %vector before the location specified by @a pos.```

From what I understand, it will copy the data from starting __first up-to (without include) of __last, in our case, `memoryLength - 1` would drop the last byte. cc @moratom for comment on the above

Additionally, if `memoryLength == 0` it would lead to undefined behaviour.

Comment (aljazkonec1) [reply src/xlink/XLinkStream.cpp:154]: Another comment from codex 😄:
 A better approach would be to resize the vector once and std::memcpy/std::copy (or even wrap the SharedMemory in a VectorMemory and hand it to the consumer) so you don’t pay the extra insert overhead.

File: src/pipeline/datatype/StreamMessageParser.cpp [hunk 6]
```diff
@@ -274,6 +275,10 @@ std::shared_ptr<ADatatype> StreamMessageParser::parseMessage(streamPacketDesc_t*
         case DatatypeEnum::ObjectTrackerConfig:
             return parseDatatype<ObjectTrackerConfig>(metadataStart, serializedObjectSize, data, fd);
             break;
+        case DatatypeEnum::PacketizedData: {
+            std::uint32_t numPackets = (std::uint32_t)readIntLE(metadataStart);
+            return std::make_shared<PacketizedData>(numPackets);
+        } break;
```
Comment (aljazkonec1) [inline src/pipeline/datatype/StreamMessageParser.cpp:300]: To keep consistency, move the logic to `parserDatatype<PacketizedData> ...`

File: tests/src/ondevice_tests/xlink_test.cpp [hunk 7]
```diff
@@ -65,4 +65,143 @@ TEST_CASE("XLinkBridge fps limit test") {
...
+
+    // Validate data integrity
+    cv::Mat returnedMat1 = returnedFrame1->getCvFrame();
+    cv::Mat returnedMat2 = returnedFrame2->getCvFrame();
+    bool same1 = std::equal(mat1.begin<uchar>(), mat1.end<uchar>(), returnedMat1.begin<uchar>());
+    bool same2 = std::equal(mat2.begin<uchar>(), mat2.end<uchar>(), returnedMat2.begin<uchar>());
+
+    REQUIRE(same1);
+    REQUIRE(same2);
+}
```
Comment (aljazkonec1) [inline tests/src/ondevice_tests/xlink_test.cpp:207]: I would add more tests, mainly for `XLinkStream` APIs, test some known byte patterns, throws / handling of empty data + non empty num packets. Maybe some tests for timeouts.

**Review by aljazkonec1 — CHANGES_REQUESTED** [review]
Thanks! Left some comments but looks good otherwise. 

There is also a compiler warning that needs to be handled:
```/home/aljaz/work-luxonis/depthai-core/src/utility/ProtoSerialize.cpp: In function ‘bool dai::utility::deserializationSupported(dai::DatatypeEnum)’:
/home/aljaz/work-luxonis/depthai-core/src/utility/ProtoSerialize.cpp:138:11: warning: enumeration value ‘PacketizedData’ not handled in switch [-Wswitch]
  138 |     switch(datatype) {
      |           ^
/home/aljaz/work-luxonis/depthai-core/src/pipeline/node/host/Replay.cpp: In function ‘std::shared_ptr<dai::Buffer> dai::node::getMessage(std::shared_ptr<google::protobuf::Message>, dai::DatatypeEnum)’:
/home/aljaz/work-luxonis/depthai-core/src/pipeline/node/host/Replay.cpp:45:11: warning: enumeration value ‘PacketizedData’ not handled in switch [-Wswitch]
   45 |     switch(datatype) {
      |           ^
/home/aljaz/work-luxonis/depthai-core/src/pipeline/node/host/Replay.cpp: In function ‘std::shared_ptr<google::protobuf::Message> dai::node::getProtoMessage(dai::utility::BytePlayer&, dai::DatatypeEnum)’:
/home/aljaz/work-luxonis/depthai-core/src/pipeline/node/host/Replay.cpp:107:11: warning: enumeration value ‘PacketizedData’ not handled in switch [-Wswitch]
  107 |     switch(datatype) {
```

File: src/pipeline/node/internal/XLinkInHost.cpp (hunk 3 above)
Comment (JakubFara) [reply src/pipeline/node/internal/XLinkInHost.cpp:53]: I added` std::uint32_t totalSize;` to [PacketizedData](https://github.com/luxonis/depthai-core/blob/feature/packetizedData/include/depthai/pipeline/datatype/PacketizedData.hpp) structure which allowed me to preallocate the vector.

File: src/pipeline/node/internal/XLinkInHost.cpp (hunk 4 above)
Comment (JakubFara) [reply src/pipeline/node/internal/XLinkInHost.cpp:73]: This change lead to some unexpected behavior.

File: src/pipeline/datatype/StreamMessageParser.cpp (hunk 6 above)
Comment (JakubFara) [reply src/pipeline/datatype/StreamMessageParser.cpp:300]: I have created [PacketizedData](https://github.com/luxonis/depthai-core/blob/feature/packetizedData/include/depthai/pipeline/datatype/PacketizedData.hpp) to be a serialized and deserelized as a regular Buffer.

File: src/pipeline/datatype/PacketizedData.hpp
Comment (JakubFara) [reply src/pipeline/datatype/PacketizedData.hpp:1]: Done here: **e8c24dd3a** _Make PacketizedData struct private_

File: include/depthai/properties/internal/XLinkOutProperties.hpp (hunk 2 above)
Comment (JakubFara) [reply include/depthai/properties/internal/XLinkOutProperties.hpp]: Changed here: **c35976900** _Change frequencyLimit to BytesPerSecondLimit_

File: include/depthai/pipeline/node/internal/XLinkOut.hpp (hunk 1 above)
Comment (JakubFara) [reply include/depthai/pipeline/node/internal/XLinkOut.hpp:55]: Yes bytes.

[I have changed ](https://github.com/luxonis/depthai-core/commit/7b2b1c56b1ce2e991f503179964421e108e918b8) `frequencyLimit` to `BytesPerSecondLimit`

File: include/depthai/properties/internal/XLinkOutProperties.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakubFara) [reply include/depthai/properties/internal/XLinkOutProperties.hpp]: fixed here **8d77a9c81** _Remove inappropriate //_

File: src/pipeline/datatype/PacketizedData.hpp [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
```
[the remark this answers is not in the document]
Comment (JakubFara) [reply src/pipeline/datatype/PacketizedData.hpp:6]: fixed

File: include/depthai/properties/internal/XLinkOutProperties.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakubFara) [reply include/depthai/properties/internal/XLinkOutProperties.hpp]: fixed here https://github.com/luxonis/depthai-core/commit/946cfd53824b7083b797f08b27ba694827390378 Remove inappropriate //

File: include/depthai/properties/internal/XLinkOutProperties.hpp [hunk 9]
```diff
@@ -18,16 +18,25 @@ struct XLinkOutProperties : PropertiesSerializable<Properties, XLinkOutPropertie
...
      */
     bool metadataOnly = false;
 
+    /**
+     * Max bytes per packet (-1 = unlimited)
+     */
+    int packetSize = -1;  //
+
+    /**
+     * Maximal frequency beteen of packets (-1 = unlimited)
```
[the remark this answers is not in the document]
Comment (JakubFara) [reply include/depthai/properties/internal/XLinkOutProperties.hpp]: fixed

File: src/pipeline/node/internal/XLinkInHost.cpp [hunk 10]
```diff
@@ -32,6 +33,42 @@ void XLinkInHost::disconnect() {
...
+    packet.data = payload.data();
+    packet.length = static_cast<uint32_t>(payload.size());
+
+    return StreamMessageParser::parseMessage(std::move(&packet));
```
[the remark this answers is not in the document]
Comment (JakubFara) [reply src/pipeline/node/internal/XLinkInHost.cpp]: fixed here: **1af49e94a**  _Remove of inappropriate usage of std::move()_

File: src/xlink/XLinkStream.cpp (hunk 5 above)
Comment (JakubFara) [reply src/xlink/XLinkStream.cpp:154]: I would leave it on a different PR?

File: tests/src/ondevice_tests/xlink_test.cpp (hunk 7 above)
Comment (JakubFara) [reply tests/src/ondevice_tests/xlink_test.cpp:207]: I you mean testing of the `class XLinkInHost` it would require mocking of the data comming from the  `std::shared_ptr<XLinkConnection>`. It is possible but it requires a good portion of time.

File: tests/CMakeLists.txt [hunk 11]
```diff
@@ -364,6 +373,10 @@ endif()
 dai_add_test(calibration_handler_test src/onhost_tests/calibration_handler_test.cpp)
 dai_set_test_labels(calibration_handler_test onhost ci)
 
+# XLInkInHost test
```
Comment (moratom) [inline tests/CMakeLists.txt:376]: XLinkInHost

File: tests/src/ondevice_tests/xlink_test.cpp (hunk 7 above)
Comment (JakubFara) [reply tests/src/ondevice_tests/xlink_test.cpp:207]: I added some tests on this commit: **53fb5c0d4** _Add tests for the XLinkInHost_

File: src/xlink/XLinkStream.cpp (hunk 5 above)
Comment (moratom) [reply src/xlink/XLinkStream.cpp:154]: >From what I understand, it will copy the data from starting __first up-to (without include) of __last, in our case, memoryLength - 1 would drop the last byte. cc @moratom for comment on the above

Should be fine, the last elemnt is at `[size-1]`

I think this code is not actually being executed now, we'll address any optimizations once zero-copy is added.

File: tests/CMakeLists.txt (hunk 11 above)
Comment (JakubFara) [reply tests/CMakeLists.txt:376]: I have added the RVC4 implementation here: https://gitlab.luxonis.com/luxonis/depthai/depthai-device-kb/-/merge_requests/246
